### PR TITLE
[Merged by Bors] - feat(Order/LatticeIntervals): order instances for Set.Icc

### PR DESCRIPTION
**Pull request by vbeffara**
Add `OrderBot`, `OrderTop` and `BoundedOrder` instances for `Set.Icc a b` under `[Fact (a \le b)]`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Order/LatticeIntervals.lean [hunk 1]
```diff
@@ -177,27 +177,41 @@ end Ici
...
+protected abbrev orderTop [Preorder α] (h : a ≤ b) : OrderTop (Icc a b) :=
   (isGreatest_Icc h).orderTop
 
+/-- `OrderTop` instance for `Icc a b`. -/
+instance [Preorder α] [Fact (a ≤ b)] : OrderTop (Icc a b) :=
+  Icc.orderTop Fact.out
+
 /-- `Icc a b` is a `BoundedOrder` whenever `a ≤ b`. -/
-protected abbrev boundedOrder [Preorder α] {a b : α} (h : a ≤ b) : BoundedOrder (Icc a b) :=
+protected abbrev boundedOrder [Preorder α] (h : a ≤ b) : BoundedOrder (Icc a b) :=
```
Comment (YaelDillies) [inline Mathlib/Order/LatticeIntervals.lean:208]: Can you instead turn the `abbrev`s into instances? Also note `Set.Icc.completeLinearOrder`

Comment (vbeffara) [reply Mathlib/Order/LatticeIntervals.lean:208]: Also `Set.Icc.completeLattice`?

**Comment by vbeffara** [comment]
@YaelDillies That ended up needing a few more changes to use the new instances. I was wandering, would an instance of `Fact (0 \le 1)` under `[ZeroLeOneClass \alpha]` be useful? E.g. it would trivialize the `BoundedOrder` instance for `UnitInterval`.

**Comment by YaelDillies** [comment]
> @YaelDillies That ended up needing a few more changes to use the new instances. I was wondering, would an instance of `Fact (0 \le 1)` under `[ZeroLeOneClass \alpha]` be useful? E.g. it would trivialize the `BoundedOrder` instance for `UnitInterval`.

It is an option, but I would keep that for a separate PR because
1. Global `Fact` instances slow down all `Fact` TC searches
2.`UnitInterval` is kind of considered a concrete type in itself

**Comment by vbeffara** [comment]
OK then I'm happy with this PR if you are.

File: Mathlib/Topology/Order/IntermediateValue.lean [hunk 2]
```diff
@@ -639,9 +639,8 @@ theorem Continuous.strictMonoOn_of_inj_rigidity {f : α → δ}
   let t := max b y
   have hsa : s ≤ a := min_le_left a x
   have hbt : b ≤ t := le_max_left b y
-  have hst : s ≤ t := hsa.trans <| hbt.trans' hab.le
   have hf_mono_st : StrictMonoOn f (Icc s t) ∨ StrictAntiOn f (Icc s t) := by
-    letI := Icc.completeLinearOrder hst
+    letI : Fact (s ≤ t) := ⟨hsa.trans <| hbt.trans' hab.le⟩
```
Comment (YaelDillies) [inline Mathlib/Topology/Order/IntermediateValue.lean:643]:
```suggestion
    have : Fact (s ≤ t) := ⟨hsa.trans <| hbt.trans' hab.le⟩
```

File: Mathlib/Topology/Order/IntermediateValue.lean [hunk 3]
```diff
@@ -664,7 +663,7 @@ theorem ContinuousOn.strictMonoOn_of_injOn_Icc {a b : α} {f : α → δ}
     (hab : a ≤ b) (hfab : f a ≤ f b)
     (hf_c : ContinuousOn f (Icc a b)) (hf_i : InjOn f (Icc a b)) :
     StrictMonoOn f (Icc a b) := by
-  letI := Icc.completeLinearOrder hab
+  letI : Fact (a ≤ b) := ⟨hab⟩
```
Comment (YaelDillies) [inline Mathlib/Topology/Order/IntermediateValue.lean:666]:
```suggestion
  have : Fact (a ≤ b) := ⟨hab⟩
```
and everywhere else

File: Mathlib/Topology/UnitInterval.lean [hunk 4]
```diff
@@ -60,7 +60,7 @@ instance hasOne : One I :=
 
 instance : ZeroLEOneClass I := ⟨zero_le_one (α := ℝ)⟩
 
-instance : BoundedOrder I := Set.Icc.boundedOrder zero_le_one
+instance : BoundedOrder I := letI : Fact ((0 : ℝ) ≤ 1) := ⟨zero_le_one⟩; inferInstance
```
Comment (YaelDillies) [inline Mathlib/Topology/UnitInterval.lean:63]:
```suggestion
instance : BoundedOrder I := have : Fact ((0 : ℝ) ≤ 1) := ⟨zero_le_one⟩; inferInstance
```

**Comment by vbeffara** [comment]
Done.

**Review by YaelDillies — APPROVED** [review]
You resolved my comments but without committing the changes, so I've redone them. All looks good to me now! Thanks for splitting

maintainer merge

**Comment by vbeffara** [comment]
Ah I missed a few, sorry

**Comment by kim-em** [comment]
bors merge

**Pull request closed without merging**